.cicd/pipeline-upload.sh

```diff
@@ -1,8 +1,24 @@
 #!/bin/bash
 set -eou pipefail
 
+# attach pipeline documentation
+echo '+++ :md: Attaching Documentation'
+export DOCS_URL="https://github.com/EOSIO/eosio.cdt/blob/${BUILDKITE_COMMIT:-master}/.cicd/README.md"
+export RETRY="$(buildkite-agent meta-data get pipeline-upload-retries --default '0')"
+buildkite-agent meta-data set pipeline-upload-retries "$(( $RETRY + 1 ))"
+if [[ "$BUILDKITE" == 'true' && "$RETRY" == '0' ]]; then
+    echo "This documentation is also available on [GitHub]($DOCS_URL)." | buildkite-agent annotate --append --style 'info' --context 'documentation'
+    cat ".cicd/README.md" | sed 's_<x>_<details>\n<summary>See More</summary>_' | sed 's_</x>_</details>_' | buildkite-agent annotate --append --style 'info' --context 'documentation'
+elif [[ "$BUILDKITE" == 'true' ]]; then
+    printf "Skipping \033]1339;url=$DOCS_URL;content=documentation\a upload for job retry number $RETRY.\n"
+fi
 export MACOS_10_14_TAG="eosio-cdt-macos-10.14-$(sha1sum ./.cicd/platforms/macos-10.14.sh | awk '{print $1}')"
 export VARS='$MACOS_10_14_TAG'
-envsubst "$VARS" < "./.cicd/pipeline.yml" > "./.cicd/pipeline.yml.out"
-buildkite-agent artifact upload ./.cicd/pipeline.yml.out
-buildkite-agent pipeline upload ./.cicd/pipeline.yml.out
+envsubst "$VARS" < ".cicd/pipeline.yml" > ".cicd/pipeline.yml.out"
+# pipeline upload
+echo '+++ :pipeline_upload: Deploying Pipeline Steps'
+if [[ "$BUILDKITE" == 'true' ]]; then
+    buildkite-agent artifact upload .cicd/pipeline.yml.out
+    buildkite-agent pipeline upload .cicd/pipeline.yml.out
+fi
```

Comment:
If we are going to use `envsubst` to explicitly call out the variables we are interpolating in the `pipeline.yml`, we should also pass `--no-interpolation` to the `buildkite-agent pipeline upload` step because it will fill in any variables we are not explicitly calling out in `envsubst`. If we are okay with `buildkite-agent` interpolating variables (I am apathetic, personally) then we should remove `envsubst` entirely. As-is, this is a misleading contradiction. It looks like the original code you are replacing was, too.

Separately, it would be great if the `pipeline.yml` we are attaching as a YAML file actually ends with the `*.yml` extension so that browsers, browser extensions, and editors are able to recognize it as YAML without having to rename the file or "tell" our tooling this is YAML.

Reply:
Yeah, its a double operation which is confusing.